pkg/controller/perconaservermongodb/users.go

```diff
@@ -0,0 +1,261 @@
+	if cr.Status.State != api.AppStateReady {
+		return nil
+	}
+
+	newSysData, err := json.Marshal(sysUsersSecretObj.Data)
+	if err != nil {
+		return errors.Wrap(err, "marshal sys secret data")
+	}
+	newSecretDataHash := sha256Hash(newSysData)
+	dataChanged, err := sysUsersSecretDataChanged(newSecretDataHash, &internalSysSecretObj)
+	if err != nil {
+		return errors.Wrap(err, "check sys users data changes")
+	}
+
+	if !dataChanged {
+		return nil
+	}
+
+	restartSfs, err := r.manageSysUsers(cr, &sysUsersSecretObj, &internalSysSecretObj)
+	if err != nil {
+		return errors.Wrap(err, "manage sys users")
+	}
+
+	internalSysSecretObj.Data = sysUsersSecretObj.Data
+	err = r.client.Update(context.TODO(), &internalSysSecretObj)
+	if err != nil {
+		return errors.Wrap(err, "update internal sys users secret")
+	}
+
+	if restartSfs {
+		err = r.restartStatefulset(cr, newSecretDataHash)
+		if err != nil {
+			return errors.Wrap(err, "restart statefulset")
+		}
+	}
+
+	return nil
+}
+
+func (r *ReconcilePerconaServerMongoDB) manageSysUsers(cr *api.PerconaServerMongoDB, sysUsersSecretObj, internalSysSecretObj *corev1.Secret) (bool, error) {
+	var (
+		sysUsers  []sysUser
+		userAdmin *sysUser
+	)
+	restartSfs := false
+	for key := range sysUsersSecretObj.Data {
+		if bytes.Compare(sysUsersSecretObj.Data[key], internalSysSecretObj.Data[key]) == 0 {
+			continue
+		}
+		switch key {
+		case "MONGODB_BACKUP_PASSWORD":
+			sysUsers = append(sysUsers, sysUser{
+				Name: string(sysUsersSecretObj.Data["MONGODB_BACKUP_USER"]),
+				Pass: string(sysUsersSecretObj.Data["MONGODB_BACKUP_PASSWORD"]),
+			},
+			)
+			restartSfs = true
+		case "MONGODB_CLUSTER_ADMIN_PASSWORD":
+			sysUsers = append(sysUsers, sysUser{
+				Name: string(sysUsersSecretObj.Data["MONGODB_CLUSTER_ADMIN_USER"]),
+				Pass: string(sysUsersSecretObj.Data["MONGODB_CLUSTER_ADMIN_PASSWORD"]),
+			},
+			)
+		case "MONGODB_CLUSTER_MONITOR_PASSWORD":
+			sysUsers = append(sysUsers, sysUser{
+				Name: string(sysUsersSecretObj.Data["MONGODB_CLUSTER_MONITOR_USER"]),
+				Pass: string(sysUsersSecretObj.Data["MONGODB_CLUSTER_MONITOR_PASSWORD"]),
+			},
+			)
+		case "MONGODB_USER_ADMIN_PASSWORD":
+			userAdmin = &sysUser{
+				Name: string(sysUsersSecretObj.Data["MONGODB_USER_ADMIN_USER"]),
+				Pass: string(sysUsersSecretObj.Data["MONGODB_USER_ADMIN_PASSWORD"]),
+			}
+		case "PMM_SERVER_PASSWORD":
+			sysUsers = append(sysUsers, sysUser{
+				Name: string(sysUsersSecretObj.Data["PMM_SERVER_USER"]),
+				Pass: string(sysUsersSecretObj.Data["PMM_SERVER_PASSWORD"]),
+			},
+			)
+			restartSfs = true
+		}
+	}
+	if userAdmin != nil {
+		sysUsers = append(sysUsers, *userAdmin)
+	}
+	if len(sysUsers) > 0 {
+		err := r.updateUsersPass(cr, sysUsers, string(internalSysSecretObj.Data["MONGODB_USER_ADMIN_USER"]), string(internalSysSecretObj.Data["MONGODB_USER_ADMIN_PASSWORD"]), internalSysSecretObj)
```

Comment:
here use "MONGODB_USER_..." strings as keys, but above - predefined consts. Do everywhere the same (which way doesn't matter)

Reply:
you forgot about consts here